src/charm.py

```diff
@@ -352,9 +353,19 @@ def _on_update_status(self, event):
             self._handle_reconfigure(event)
 
         # update the units status based on it's replica set status.
-        with MongoDBConnection(self.mongodb_config) as mongo:
-            replset_status = mongo.get_replset_status()
-            self.unit.status = build_unit_status(replset_status, self._unit_ip(self.unit))
+        try:
+            with MongoDBConnection(self.mongodb_config) as mongo:
+                replset_status = mongo.get_replset_status()
+                self.unit.status = build_unit_status(replset_status, self._unit_ip(self.unit))
+        except ServerSelectionTimeoutError as e:
+            # ServerSelectionTimeoutError is commonly due to ReplicaSetNoPrimary
+            logger.debug("Got error: %s, while checking replica set status", str(e))
+            self.unit.status = WaitingStatus("Waiting for primary re-election..")
+        except AutoReconnect as e:
+            # AutoReconnect is raised when a connection to the database is lost and an attempt to
+            # auto-reconnect will be made by pymongo.
+            logger.debug("Got error: %s, while checking replica set status", str(e))
+            self.unit.status = WaitingStatus("Waiting to reconnect to unit..")
```

Comment:
I know it's another comment in the same place, but as you're not using `replset_status` anywhere else in this method, could you modify `build_unit_status` to take `mongodb_config` as an arg instead of `replset_status`, build the status there, and do the context-manager `with`, and the error handling there too?

Reply:
good idea :)